src/tools/cgexec.c

```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+			} else if (ret == 2) {
+				found_unified_cgrp = 1;
+				continue;
+			}
+		}
+
+		/* skip if the cgroup path doesn't have systemd scope format */
+		if (strstr(cgroup_name, ".scope") == NULL ||
+		    strstr(cgroup_name, ".slice") == NULL)
+			continue;
+
+		/* skip if we have already searched cgroup for idle_thread */
+		if (is_scope_parsed(cgroup_name))
+			continue;
+
+		/* cgroup v1 might have shared mount points cpu,cpuacct */
+		_ctrl_name = strchr(ctrl_name, ',');
+		if (_ctrl_name) {
+			size = strlen(ctrl_name) - strlen(_ctrl_name);
+			ctrl_name[size] = '\0';
+		}
+
+		if (ret == 2)
+			ret = cgroup_get_procs(cgroup_name, NULL, &pids, &size);
+		else
+			ret = cgroup_get_procs(cgroup_name, ctrl_name, &pids, &size);
+		if (ret) {
+			err("Failed to read cgroup.procs of cgroup: %s\n", cgroup_name + 1);
+			goto out;
+		}
+
+		/*
+		 * .scope created by the non-libcgroup process, will not
+		 * have libcgroup_systemd_idle_thread
+		 */
+		_scope_pid = search_systemd_idle_thread_task(pids, size);
+		if (_scope_pid == -1)
+			continue;
+
+		if (scope_pid == -1) {
+			/*
+			 * cgexec pid needs to written into:
+			 * ../systemd/<slice>/<scope>/cgroup.procs (legacy/hybrid)
+			 * ../unified/<slice>/<scope>/cgroup.procs (hybrid)
+			 */
+			snprintf(systemd_cgrp_name, FILENAME_MAX, "%s", cgroup_name);
+			snprintf(systemd_ctrl_name, FILENAME_MAX, "%s", ctrl_name);
+			snprintf(unified_cgrp_name, FILENAME_MAX, "%s", cgroup_name);
+			scope_pid = _scope_pid;
+			continue;
+		}
+
+		if (_scope_pid != scope_pid) {
+			err("Failed to replace scope idle_thread, found two idle_thread\n");
+			err(" %u %u\n", scope_pid, _scope_pid);
+			goto out;
+		}
+	}
+
+	if (scope_pid == -1) {
+		err("Failed to find idle_thread task\n");
+		goto out;
+	}
+
+	if (is_cgroup_mode_legacy() && (found_systemd_cgrp == 0 || found_unified_cgrp == 1)) {
+			err("cgroup legacy setup incorrect\n");
+			scope_pid = -1;
+			goto out;
+	}
+
+	if (is_cgroup_mode_hybrid() && (found_systemd_cgrp == 0 || found_unified_cgrp == 0)) {
+			err("cgroup hybrid setup incorrect\n");
+			scope_pid = -1;
+			goto out;
+	}
+
+	/* This is true for cgroup v1 (legacy/hybrid) */
+	if (found_systemd_cgrp) {
+		ret = write_systemd(systemd_cgrp_name, systemd_ctrl_name, unified_cgrp_name);
+		if (ret)
+			scope_pid = -1;
+	}
+
+out:
+	if (pid_proc_fp)
+		fclose(pid_proc_fp);
+
+	return scope_pid;
+}
+
+static int write_systemd(const char * const systemd_cgrp_name,
+			 const char * const systemd_ctrl_name,
+			 const char * const unified_cgrp_name)
+{
+	char cgroup_procs_path[FILENAME_MAX * 2 + 25];
+	char systemd_cgroup_name[FILENAME_MAX];
+	FILE *cgroup_systemd_path_f = NULL;
+	FILE *cgroup_unified_path_f = NULL;
+	char *ctrl_name = NULL;
+	pid_t pid;
+	int len;
+
+	pthread_rwlock_rdlock(&cg_mount_table_lock);
+	cg_build_path_locked("", systemd_cgroup_name, systemd_ctrl_name);
+	pthread_rwlock_unlock(&cg_mount_table_lock);
+
+	ctrl_name = strstr(systemd_cgroup_name, systemd_ctrl_name);
```

Comment:
It might be good to add a couple comments of what you expect the string to look like each step of the way here.  I won't remember when I re-look at this in a few months (or days) :)

Reply:
Agreed, will add more comments.